### PR TITLE
Fix part of #6106: Fixes issues in play console run

### PR DESCRIPTION
**Pull request by Sandesh282**
Fixes part of #6106

## Explanation

This PR fixes several issues in the Play Console deployment pipeline that were preventing successful releases:

1. WIF-based Authentication (all deployment workflows)
All three deployment workflows (deploy_to_play_console.yml, deploy_updated_changelog.yml, update_rollout.yml) previously obtained access tokens via gcloud auth print-access-token, which doesn't work with Workload Identity Federation (WIF). They now use google-github-actions/auth@v2 with token_format: 'access_token' to obtain a proper OAuth2 access token directly from the WIF exchange. The token is masked before use to prevent accidental log exposure.

2. Version Inversion Checker — Edit Session Reuse
VersionInversionChecker previously created temporary edit sessions to read track releases, then deleted them before the upload edit was opened. The Play Developer API enforces a single active edit per package, so this created a race window. The checker now receives the already-open upload editId from UploadBinaryToPlayConsole and reuses it for all three track queries. The deleteEdit method has been removed from the interface, implementation, and all tests since it is no longer needed.

3. deploy_updated_changelog.yml — Correct Access Token
The script invocation in this workflow was using gcloud auth print-access-token while the auth step uses WIF and does not configure gcloud credentials. Fixed to use the access token output directly from the WIF auth step.

4. build_and_sign.yml — `zip -d` Error Handling
The || true after zip -d silently swallowed all failures (corrupt archive, missing binary, permission errors), which could allow a still-multi-signed AAB to reach Play Console. Fixed: exit code 12 ("no matching entries" — AAB was already clean) is treated as non-fatal; all other non-zero exit codes stop the workflow.

5. deploy_to_play_console.yml — Cleanup
Removed the unnecessary `--action_env=GOOGLE_APPLICATION_CREDENTIALS` Bazel flag (not needed when passing the WIF access token directly) and an unnecessary inline comment documenting script argument names.

Note: Parts of the PR description were generated with the help of Claude Code.
## Essential Checklist
- [x] The PR title starts with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The explanation section above starts with "Fixes #bugnum: " (If this PR fixes part of an issue, use instead: "Fixes part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved Play Console uploads by removing conflicting signing metadata from app bundles.
  * Updated Google Play API routing and authentication to improve deployment reliability.
  * Improved staged rollout updates using the required access permissions.

### CODE REVIEW

File: .github/workflows/deploy_to_play_console.yml
Comment (BenHenning) [inline .github/workflows/deploy_to_play_console.yml]: This can be removed now.

File: .github/workflows/build_and_sign.yml [hunk 1]
```diff
@@ -228,5 +228,12 @@ jobs:
             exit 1
           fi
 
+          # Strip jarsigner-appended META-INF signature files before upload.
+          # jarsigner appends new signing certs rather than replacing the originals, leaving
+          # the AAB multi-signed. Play Console rejects multi-signed AABs, so we remove the
+          # old signatures here before archiving. apksigner (used at deploy time) will replace
+          # these with the upload certificate during signing.
+          zip -d "$SIGNED_AAB_PATH" "META-INF/*.RSA" "META-INF/*.SF" "META-INF/*.DSA" || true
+
```
[the remark this answers is not in the document]
Comment (BenHenning) [reply .github/workflows/build_and_sign.yml:237]: This seems like good feedback--the workflow should fail if this step fails otherwise we might not catch a problem until much later (trying to upload the AAB).

File: .github/workflows/deploy_to_play_console.yml [hunk 2]
```diff
@@ -109,6 +118,7 @@ jobs:
 
       # Runs the UploadBinaryToPlayConsole Kotlin script which enforces all precondition checks
       # (pending release, version inversion, changelog existence) before uploading the AAB.
+      # Script args: workspaceRoot, aabPath, track, accessToken, rolloutFraction
```
Comment (BenHenning) [inline .github/workflows/deploy_to_play_console.yml]: Not necessary (we don't do this for other scripts)--please remove.

File: scripts/src/java/org/oppia/android/scripts/release/remote/PlayConsoleService.kt [hunk 3]
```diff
@@ -33,7 +33,7 @@ interface PlayConsoleService {
    * @param authorizationBearer the OAuth2 bearer token for authentication
    * @return the [EditResponse] containing the new edit session ID
    */
-  @POST("{packageName}/edits")
+  @POST("androidpublisher/v3/applications/{packageName}/edits")
```
[the remark this answers is not in the document]
Comment (BenHenning) [reply scripts/src/java/org/oppia/android/scripts/release/remote/PlayConsoleService.kt:36]: Seems worth fixing--I assumed tests would need updating as well. @Sandesh282

File: scripts/src/java/org/oppia/android/scripts/release/GooglePlayConsoleClient.kt [hunk 4]
```diff
@@ -159,15 +160,24 @@ class GooglePlayConsoleClient(
     }
   }
 
+  override fun deleteEdit(packageName: String, editId: String) {
+    val response = playConsoleService
+      .deleteEdit(packageName, editId, authorizationBearer)
```
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/release/GooglePlayConsoleClient.kt]: Why are we adding deletion support? As far as I can tell we aren't using it anywhere except tests, and it shouldn't be necessary. Am I missing something?

File: scripts/src/java/org/oppia/android/scripts/release/PlayConsoleClient.kt [hunk 5]
```diff
@@ -26,7 +26,11 @@ interface PlayConsoleClient {
    * @param track the Play Console track to query (e.g. "alpha", "beta", "production")
    * @return the list of [TrackRelease] entries currently live on the track
    */
-  fun getTrackReleases(packageName: String, track: String): List<TrackRelease>
+  fun getTrackReleases(
+    packageName: String,
+    track: String,
+    existingEditId: String? = null
```
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/release/PlayConsoleClient.kt:37]: Here & elsewhere: update KDocs & related tests (for each case of the new `existingEditId` parameter.

File: .github/workflows/deploy_to_play_console.yml
Comment (Sandesh282) [reply .github/workflows/deploy_to_play_console.yml]: Done.

File: .github/workflows/deploy_to_play_console.yml (hunk 2 above)
Comment (Sandesh282) [reply .github/workflows/deploy_to_play_console.yml]: Removed.

File: scripts/src/java/org/oppia/android/scripts/release/GooglePlayConsoleClient.kt (hunk 4 above)
Comment (Sandesh282) [reply scripts/src/java/org/oppia/android/scripts/release/GooglePlayConsoleClient.kt]: Removed.

File: scripts/src/java/org/oppia/android/scripts/release/PlayConsoleClient.kt (hunk 5 above)
Comment (Sandesh282) [reply scripts/src/java/org/oppia/android/scripts/release/PlayConsoleClient.kt:37]: Updated in PlayConsoleClient.kt, GooglePlayConsoleClient.kt, FakePlayConsoleClient.kt

File: .github/workflows/build_and_sign.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (BenHenning) [reply .github/workflows/build_and_sign.yml:237]: Why the `||` here? Why can't we just let `zip` fail the workflow?

File: scripts/src/java/org/oppia/android/scripts/release/remote/PlayConsoleService.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (Sandesh282) [reply scripts/src/java/org/oppia/android/scripts/release/remote/PlayConsoleService.kt:36]: @BenHenning I was thinking of skipping it since it's not a critical/major issue.

File: .github/workflows/deploy_updated_changelog.yml [hunk 6]
```diff
@@ -184,4 +184,4 @@ jobs:
             "$GITHUB_WORKSPACE" \
             "org.oppia.android" \
             "$VERSION" \
-            "$(gcloud auth print-access-token)"
+            "${{ steps.auth.outputs.access_token }}"
```
Comment (BenHenning) [inline .github/workflows/deploy_updated_changelog.yml]: Won't this also require the token scopes?

File: .github/workflows/build_and_sign.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (Sandesh282) [reply .github/workflows/build_and_sign.yml:237]: Done.

File: scripts/src/java/org/oppia/android/scripts/release/VersionInversionChecker.kt [hunk 7]
```diff
@@ -32,12 +32,17 @@ class VersionInversionChecker(private val client: PlayConsoleClient) {
    * @param newVersionCode the version code of the binary about to be uploaded
    * @throws IllegalStateException if [newVersionCode] violates the cross-track ordering constraint
    */
-  fun verify(packageName: String, targetTrack: String, newVersionCode: Long) {
-    val alphaVersionCodes = client.getTrackReleases(packageName, ALPHA_TRACK)
+  fun verify(
+    packageName: String,
+    targetTrack: String,
+    newVersionCode: Long,
+    existingEditId: String
```
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/release/VersionInversionChecker.kt:42]: Still missing updated KDoc for this.

File: scripts/src/javatests/org/oppia/android/scripts/release/GooglePlayConsoleClientTest.kt [hunk 8]
```diff
@@ -223,7 +223,7 @@ class GooglePlayConsoleClientTest {
 
   @Test
   fun testGetTrackReleases_parsesReleasesFromResponse() {
-    // getTrackReleases internally calls createEdit first.
+    // getTrackReleases internally calls createEdit then getTrack.
```
Comment (BenHenning) [inline scripts/src/javatests/org/oppia/android/scripts/release/GooglePlayConsoleClientTest.kt]: Seems like an unnecessary comment?

File: scripts/src/java/org/oppia/android/scripts/release/VersionInversionChecker.kt (hunk 7 above)
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/release/VersionInversionChecker.kt:42]: We should have a target test to make sure this is actually doing the correct thing (in the context of `VersionInversionChecker`).

File: scripts/src/java/org/oppia/android/scripts/release/GooglePlayConsoleClient.kt [hunk 9]
```diff
@@ -80,10 +80,12 @@ class GooglePlayConsoleClient(
...
     packageName: String,
-    track: String
+    track: String,
+    existingEditId: String?
   ): List<PlayConsoleClient.TrackRelease> {
-    // A temporary edit is needed to read track info via the API.
-    val editId = createEdit(packageName)
+    // Use the caller's open edit when provided to avoid creating a new temporary edit session.
+    // If no edit ID is supplied, create a temporary one just for this read.
+    val editId = existingEditId ?: createEdit(packageName)
```
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/release/GooglePlayConsoleClient.kt:88]: We should have a target test to make sure this is actually doing the correct thing (in the context of `GooglePlayConsoleClient`). This one is particularly important because it's validating that the logic is actually correct for the core issue that caused the script to fail.

File: scripts/src/java/org/oppia/android/scripts/release/UploadBinaryToPlayConsole.kt [hunk 10]
```diff
@@ -115,7 +115,7 @@ private fun runUpload(
   println("Uploaded — assigned version code: $uploadedVersionCode")
 
   println("Running version inversion check...")
-  VersionInversionChecker(client).verify(PACKAGE_NAME, track, uploadedVersionCode)
+  VersionInversionChecker(client).verify(PACKAGE_NAME, track, uploadedVersionCode, editId)
```
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/release/UploadBinaryToPlayConsole.kt:118]: We should have a target test to make sure this is actually doing the correct thing (in the context of `UploadBinaryToPlayConsole`).

**Review by BenHenning — CHANGES_REQUESTED** [review]
Took a full pass, PTAL @Sandesh282.

File: scripts/src/java/org/oppia/android/scripts/release/remote/PlayConsoleService.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (BenHenning) [reply scripts/src/java/org/oppia/android/scripts/release/remote/PlayConsoleService.kt:36]: Won't `testInsertEdit_sendsPostToCorrectPath` fail because of this not being updated? Have you double checked that all of your script tests are passing locally before pushing the PR @Sandesh282?

File: .github/workflows/build_and_sign.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (BenHenning) [reply .github/workflows/build_and_sign.yml:237]: FYI my comment came in before yours @Sandesh282 but it's not addressed (since you posted that mid-review, the comment timing is a bit goofed).

File: scripts/src/java/org/oppia/android/scripts/release/remote/PlayConsoleService.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (Sandesh282) [reply scripts/src/java/org/oppia/android/scripts/release/remote/PlayConsoleService.kt:36]: @BenHenning Yes, I have tested all my scripts. I didn't face this issue. But I will double check and add this fix if it's important.

PS: I verified by running the test again. I already fixed it [here](https://github.com/oppia/oppia-android/pull/6328/commits/c0e200dc2b9a3be650a7495dce0680e7f6e3a79d).
image

File: .github/workflows/deploy_updated_changelog.yml (hunk 6 above)
Comment (Sandesh282) [reply .github/workflows/deploy_updated_changelog.yml]: Added id: auth, token_format: 'access_token', access_token_scopes, and masking step

File: scripts/src/java/org/oppia/android/scripts/release/VersionInversionChecker.kt (hunk 7 above)
Comment (Sandesh282) [reply scripts/src/java/org/oppia/android/scripts/release/VersionInversionChecker.kt:42]: Updated.

File: scripts/src/javatests/org/oppia/android/scripts/release/GooglePlayConsoleClientTest.kt (hunk 8 above)
Comment (Sandesh282) [reply scripts/src/javatests/org/oppia/android/scripts/release/GooglePlayConsoleClientTest.kt]: Removed.

File: scripts/src/java/org/oppia/android/scripts/release/VersionInversionChecker.kt (hunk 7 above)
Comment (Sandesh282) [reply scripts/src/java/org/oppia/android/scripts/release/VersionInversionChecker.kt:42]: Added testVerify_existingEditId_isPassedToAllThreeTrackQueries.

File: scripts/src/java/org/oppia/android/scripts/release/GooglePlayConsoleClient.kt (hunk 9 above)
Comment (Sandesh282) [reply scripts/src/java/org/oppia/android/scripts/release/GooglePlayConsoleClient.kt:88]: Added testGetTrackReleases_withExistingEditId_reusesEditWithoutCreatingNew + _withoutExistingEditId_createsNewEditFirst

File: scripts/src/java/org/oppia/android/scripts/release/UploadBinaryToPlayConsole.kt (hunk 10 above)
Comment (Sandesh282) [reply scripts/src/java/org/oppia/android/scripts/release/UploadBinaryToPlayConsole.kt:118]: Added testRunUpload_versionInversionChecker_reusesUploadEditIdForAllThreeTrackQueries

File: .github/workflows/build_and_sign.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (Sandesh282) [reply .github/workflows/build_and_sign.yml:237]: @BenHenning I have addressed it now.

File: scripts/src/javatests/org/oppia/android/scripts/release/GooglePlayConsoleClientTest.kt [hunk 11]
```diff
@@ -261,4 +260,43 @@ class GooglePlayConsoleClientTest {
 
     assertThat(exception).hasMessageThat().contains("404")
   }
+  @Test
```
Comment (BenHenning) [inline scripts/src/javatests/org/oppia/android/scripts/release/GooglePlayConsoleClientTest.kt:264]: Nit: add newline before this for consistent spacing.

File: scripts/src/javatests/org/oppia/android/scripts/release/VersionInversionCheckerTest.kt [hunk 12]
```diff
@@ -295,9 +400,56 @@ class VersionInversionCheckerTest {
...
+        "org.oppia.android",
+        "internal",
+        newVersionCode = 100L,
+        existingEditId = "test-edit"
+      )
     }
 
     assertThat(exception).hasMessageThat().contains("internal")
   }
+  // ---------------------------------------------------------------------------
```
Comment (BenHenning) [inline scripts/src/javatests/org/oppia/android/scripts/release/VersionInversionCheckerTest.kt]: Newline before here as well, though I think we can remove these line comments since it's just repeating what the test name already indicates.

File: scripts/src/javatests/org/oppia/android/scripts/release/VersionInversionCheckerTest.kt [hunk 13]
```diff
@@ -295,9 +400,56 @@ class VersionInversionCheckerTest {
...
+  // existingEditId — passed through to getTrackReleases on all three tracks
+  // ---------------------------------------------------------------------------
+
+  @Test
+  fun testVerify_existingEditId_isPassedToAllThreeTrackQueries() {
+    // The FakePlayConsoleClient records the existingEditId received by getTrackReleases via
+    // the last queried editId. We use a custom fake to capture the IDs.
+    val capturedEditIds = mutableListOf<String?>()
+    val capturingFake = object : PlayConsoleClient {
```
Comment (BenHenning) [inline scripts/src/javatests/org/oppia/android/scripts/release/VersionInversionCheckerTest.kt]: Can't this be observed via the `FakePlayConsoleClient` already being used to indirectly configure the version checker? It seems a bit counter to me to use a different, custom fake when there's already one available for observing `PlayConsoleClient` behaviors.

**Review by BenHenning — COMMENTED** [review]
Thanks @Sandesh282! Just a few follow-ups, PTAL.

File: scripts/src/javatests/org/oppia/android/scripts/release/GooglePlayConsoleClientTest.kt (hunk 11 above)
Comment (Sandesh282) [reply scripts/src/javatests/org/oppia/android/scripts/release/GooglePlayConsoleClientTest.kt:264]: Done.

File: scripts/src/javatests/org/oppia/android/scripts/release/VersionInversionCheckerTest.kt (hunk 12 above)
Comment (Sandesh282) [reply scripts/src/javatests/org/oppia/android/scripts/release/VersionInversionCheckerTest.kt]: Done.

File: scripts/src/javatests/org/oppia/android/scripts/release/VersionInversionCheckerTest.kt (hunk 13 above)
Comment (Sandesh282) [reply scripts/src/javatests/org/oppia/android/scripts/release/VersionInversionCheckerTest.kt]: Updated.

**Comment by Sandesh282** [comment]
@BenHenning PTAL!

File: scripts/src/java/org/oppia/android/scripts/release/FakePlayConsoleClient.kt [hunk 14]
```diff
@@ -30,6 +31,12 @@ class FakePlayConsoleClient : PlayConsoleClient {
   /** All edit session IDs committed via [commitEdit], in order. */
   val committedEdits = mutableListOf<String>()
 
+  /**
+   * The [existingEditId] argument passed to each [getTrackReleases] call, in order.
+   * `null` entries indicate calls where no existing edit was provided.
+   */
+  val queriedEditIds = mutableListOf<String?>()
```
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/release/FakePlayConsoleClient.kt:38]: Please also add `FakePlayConsoleClientTest` tests for making this is empty, populated, and cleared when expected. We generally should always be adding tests for all new functionality in each class (it's rare for us to not do that). I suggest preempting this in the future by always making sure updated functionality is always tested.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @Sandesh282. Just one comment left, PTAL.

File: scripts/src/java/org/oppia/android/scripts/release/FakePlayConsoleClient.kt (hunk 14 above)
Comment (Sandesh282) [reply scripts/src/java/org/oppia/android/scripts/release/FakePlayConsoleClient.kt:38]: Done.

**Review by BenHenning — APPROVED** [review]
Thanks @Sandesh282. This LGTM.

**Comment by BenHenning** [comment]
Enabling auto-merge.
